vita3k/interface.cpp

```diff
@@ -272,27 +273,75 @@ static ExitCode load_app_impl(Ptr<const void> &entry_point, HostState &host, Gui
 
     host.renderer->features.hardware_flip = host.cfg.hardware_flip;
 
-    // Load pre-loaded libraries
-    const char *const lib_load_list[] = {
-        "sce_module/libc.suprx",
-        "sce_module/libfios2.suprx",
-        "sce_module/libult.suprx",
+    // Load user pre-loaded module
+    const char *const lib_load_os0_list[] = {
+        "us/driver_us.suprx",
     };
 
-    for (auto module_path : lib_load_list) {
+    for (auto module_path : lib_load_os0_list) {
         vfs::FileBuffer module_buffer;
         Ptr<const void> lib_entry_point;
 
-        if (vfs::read_app_file(module_buffer, host.pref_path, host.io.title_id, module_path)) {
-            SceUID module_id = load_self(lib_entry_point, host.kernel, host.mem, module_buffer.data(), std::string("app0:") + module_path, host.cfg);
+        if (vfs::read_file(VitaIoDevice::os0, module_buffer, host.pref_path, module_path)) {
+            SceUID module_id = load_self(lib_entry_point, host.kernel, host.mem, module_buffer.data(), std::string("os0:") + module_path, host.cfg);
             if (module_id >= 0) {
                 const auto module = host.kernel.loaded_modules[module_id];
 
-                LOG_INFO("Pre-load module {} (at \"{}\") loaded", module->module_name, module_path);
+                LOG_INFO("User pre-load module {} (at \"{}\") loaded", module->module_name, module_path);
             } else
                 return FileNotFound;
         } else {
-            LOG_DEBUG("Pre-load module at \"{}\" not present", module_path);
+            LOG_DEBUG("User Pre-load module at \"{}\" not present", module_path);
+        }
+    }
+
+    // Load pre-loaded libraries
+    const auto module_app_path{ fs::path(host.pref_path) / "ux0/app" / host.io.title_id / "sce_module" };
```

Comment:
Why not just create a temp variable and std::string(temp) + module_path to eliminate those duplicated code.

Reply:
sorry, no understund what you want ask 🤔

Reply:
```cpp

 if (fs::exists(module_app_path) && !fs::is_empty(module_app_path)) {
   device= VitaIoDevice::app0;
   prefix = std::string("app0:");
} else {
  device = VitaIoDevice::vs0;
  prefix = std::string("vs0:");
}
        const char *const lib_load_list[] = {
            "sce_module/libc.suprx",
            "sce_module/libfios2.suprx",
            "sce_module/libult.suprx",
        };

        for (auto &module_path : lib_load_list) {
            vfs::FileBuffer module_buffer;
            Ptr<const void> lib_entry_point;

            if (vfs::read_file(device, module_buffer, host.pref_path, host.io.title_id, module_path)) {
                SceUID module_id = load_self(lib_entry_point, host.kernel, host.mem, module_buffer.data(), prefix + module_path, host.cfg);
                if (module_id >= 0) {
                    const auto module = host.kernel.loaded_modules[module_id];

                    LOG_INFO("Pre-load module {} (at \"{}\") loaded", module->module_name, module_path);
                } else
                    return FileNotFound;
            } else {
                LOG_DEBUG("Pre-load module at \"{}\" not present", module_path);
            }
        }

```

Reply:
just a pseudocode no able to use computer

Reply:
look more, is not possible like that, path for module is diferent

Reply:
then you can also set lib_load_list before

Reply:
path module is diferent i said, i can't like that

Reply:
I meant

```
if (fs::exists(module_app_path) && !fs::is_empty(module_app_path)) {
   device= VitaIoDevice::app0;
   prefix = std::string("app0:");
const char *const lib_load_list[] = {
            "sce_module/libc.suprx",
            "sce_module/libfios2.suprx",
            "sce_module/libult.suprx",
        };

} else {
  device = VitaIoDevice::vs0;
  prefix = std::string("vs0:");
 const char *const lib_load_list[] = {
            "sys/external/libc.suprx",
            "sys/external/libfios2.suprx",
            "sys/external/libult.suprx",
        };
}

        for (auto &module_path : lib_load_list) {
            vfs::FileBuffer module_buffer;
            Ptr<const void> lib_entry_point;

            if (vfs::read_file(device, module_buffer, host.pref_path, host.io.title_id, module_path)) {
                SceUID module_id = load_self(lib_entry_point, host.kernel, host.mem, module_buffer.data(), prefix + module_path, host.cfg);
                if (module_id >= 0) {
                    const auto module = host.kernel.loaded_modules[module_id];

                    LOG_INFO("Pre-load module {} (at \"{}\") loaded", module->module_name, module_path);
                } else
                    return FileNotFound;
            } else {
                LOG_DEBUG("Pre-load module at \"{}\" not present", module_path);
            }
        }
```

Reply:
AHH yes, i have n,o thinks like that but your code no can works here, need inisalise it before if lol

Reply:
you can make it work if you tweak it little, right? I believe in you. 😎

Reply:
check again, i have reworks it, with diferent approch, and adios this char*...  i hate char xD

Reply:
looks great to me. btw we've decided to not load driver_user by default.

Reply:
I'm not sure whether it will cause regressions or not even. Can you test loading it with many games.

Reply:
ah, drop, forget this smart idea, is not possible, obligatory need use read app file and read file separed sorry :/  here is impossible read module inside game lol

Reply:
```
bool is_app = fs::exists(module_app_path) && !fs::is_empty(module_app_path);
 if (is_app) {
   prefix = std::string("app0:");
        const char *const lib_load_list[] = {
            "sce_module/libc.suprx",
            "sce_module/libfios2.suprx",
            "sce_module/libult.suprx",
        };

} else {
  prefix = std::string("vs0:");
 const char *const lib_load_list[] = {
            "sys/external/libc.suprx",
            "sys/external/libfios2.suprx",
            "sys/external/libult.suprx",
        };
}

        for (auto &module_path : lib_load_list) {
            vfs::FileBuffer module_buffer;
            Ptr<const void> lib_entry_point;
            bool error;
             if (is_app)
       error = vfs::read_app_file(module_buffer, host.pref_path, host.io.title_id, module_path);
            else
  error = vfs::read_file(VitaIoDevice::vs0, module_buffer, host.pref_path, host.io.title_id, module_path);
        
            if (error) {
                SceUID module_id = load_self(lib_entry_point, host.kernel, host.mem, module_buffer.data(), prefix + module_path, host.cfg);
                if (module_id >= 0) {
                    const auto module = host.kernel.loaded_modules[module_id];
                    LOG_INFO("Pre-load module {} (at \"{}\") loaded", module->module_name, module_path);
                } else
                    return FileNotFound;
            } else {
                LOG_DEBUG("Pre-load module at \"{}\" not present", module_path);
            }
        }
```

Reply:
no very like this view, drafted i see later